### ISSUE TITLE
Fail Gracefully with NULL tensors

### ISSUE THREAD

**Issue by jonthegeek** [mlverse/torch#837]
NULLs passed to torch_tensor fail gracefully.
``` r
torch::torch_tensor(NULL)
#> Error in torch_tensor_cpp(data, dtype, device, requires_grad, pin_memory): R type not handled
```

NULLs passed to torch_tensor as integers "work".
```r
torch::torch_tensor(as.integer(NULL))
#> torch_tensor
#> [ CPULongType{0} ]
```

But if you subset that tensor, it causes R to crash.
```r
# Crashes Rstudio.
torch::torch_tensor(as.integer(NULL))[1]
```

This can happen relatively easily with a malformed dataset (for example, if you try to construct labels in a test set), and the crash is mysterious.

Should `torch_tensor` validate the `data` argument to give a friendlier error message for (any type of) NULL?

**Comment by dfalbel** [comment]
I think I'd split this into 2 issues:

- What happens when we create tensors with `torch::torch_tensor(NULL)`?

We should probably raise a better error message stating what's expected and why `NULL` is not allowed.

- What happens when we create a tensor from a atomic vector with length 0?

In this case I think it makes sense to be able to create the tensor (instead of failing) as tensors with 0 elements are allowed in torch and it's ok if they happen to exist.

However, subsetting a length 0 tensor should raise an error instead of crashing the R session. That's certainly a bug.

What do you think?

**Comment by dfalbel** [comment]
There's should be no difference between length-0 tensors and tensors created with `torch_tensor(as.integer(NULL))`, that's because: `as.integer(NULL)` is the same as `integer(0)`.

Maybe we should default `torch_tensor(NULL)` to create a lenght 0 tensor with a default dtype, eg. float32., so it would be equivalent to `torch_tensor(numeric(0))`.

**Comment by jonthegeek** [comment]
OHHHH, right! Ha, in my mind it was just NULL, I hadn't ran it to realize what actually happens. Which duh, ok, now it's obvious :D

**Comment by dfalbel** [comment]
I think #841 will fix the crash! Please feel free to open a new issue if it doesn't work.
